pkg/kubelet/pluginmanager/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,267 @@
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"k8s.io/klog"
+
+	"k8s.io/kubernetes/pkg/kubelet/pluginmanager/cache"
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
+)
+
+// Watcher is the plugin watcher
+type Watcher struct {
+	path                string
+	deprecatedPath      string
+	fs                  utilfs.Filesystem
+	fsWatcher           *fsnotify.Watcher
+	wg                  sync.WaitGroup
+	desiredStateOfWorld cache.DesiredStateOfWorld
+}
+
+// NewWatcher provides a new watcher
+// deprecatedSockDir refers to a pre-GA directory that was used by older plugins
+// for socket registration. New plugins should not use this directory.
+func NewWatcher(sockDir string, deprecatedSockDir string, desiredStateOfWorld cache.DesiredStateOfWorld) *Watcher {
+	return &Watcher{
+		path:                sockDir,
+		deprecatedPath:      deprecatedSockDir,
+		fs:                  &utilfs.DefaultFs{},
+		desiredStateOfWorld: desiredStateOfWorld,
+	}
+}
+
+// Start watches for the creation and deletion of plugin sockets at the path
+func (w *Watcher) Start(stopCh <-chan struct{}) error {
+	klog.V(2).Infof("Plugin Watcher Start at %s", w.path)
+
+	// Creating the directory to be watched if it doesn't exist yet,
+	// and walks through the directory to discover the existing plugins.
+	if err := w.init(); err != nil {
+		return err
+	}
+
+	fsWatcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to start plugin fsWatcher, err: %v", err)
+	}
+	w.fsWatcher = fsWatcher
+
+	w.wg.Add(1)
+	go func(fsWatcher *fsnotify.Watcher) {
+		defer w.wg.Done()
+		for {
+			select {
+			case event := <-fsWatcher.Events:
+				//TODO: Handle errors by taking corrective measures
+
+				w.wg.Add(1)
+				func() {
+					defer w.wg.Done()
+					if event.Op&fsnotify.Create == fsnotify.Create {
+						err := w.handleCreateEvent(event)
+						if err != nil {
+							klog.Errorf("error %v when handling create event: %s", err, event)
+						}
+					} else if event.Op&fsnotify.Remove == fsnotify.Remove {
+						err := w.handleDeleteEvent(event)
+						if err != nil {
+							klog.Errorf("error %v when handling delete event: %s", err, event)
+						}
+					}
+					return
+				}()
+				continue
+			case err := <-fsWatcher.Errors:
+				if err != nil {
+					klog.Errorf("fsWatcher received error: %v", err)
+				}
+				continue
+			case <-stopCh:
+				// In case of plugin watcher being stopped by plugin manager, stop
+				// probing the creation/deletion of plugin sockets
+				c := make(chan struct{})
+				go func() {
+					defer close(c)
+					w.wg.Wait()
+				}()
+				select {
+				case <-c:
+				case <-time.After(11 * time.Second):
+					klog.Errorf("timeout on stopping watcher")
+				}
+				w.fsWatcher.Close()
+				return
+			}
+		}
+	}(fsWatcher)
+
+	// Traverse plugin dir after starting the plugin processing goroutine
+	if err := w.traversePluginDir(w.path); err != nil {
+		klog.Errorf("failed to traverse plugin socket path %q, err: %v", w.path, err)
```

Comment:
Previously for both `w.traversePluginDir(w.path)` and `w.traversePluginDir(w.deprecatedPath)` an error would be terminal:

https://github.com/kubernetes/kubernetes/blob/28172ec17c088527bf543e65d2a2019ba7add20a/pkg/kubelet/util/pluginwatcher/plugin_watcher.go#L142-L154

Looks like you changed it to log and continue? I want to make sure I understand the reason behind the change.

Reply:
@saad-ali thanks for catching this. I'm actually a little unsure about my change. Currently, kubelet is in charge of starting the plugin watcher, and not in charge of stopping it. In this PR, plugin manager will now in charge of starting and stopping the watcher. So now, the kubelet will start the plugin manager, and plugin manager will start/stop the reconciler and the plugin watcher. 

I think this is better than stopping the watcher if an error is encountered when traversing the directories because I don't see a way for the kubelet to restart the watcher again. Does it make sense for the watcher to stop if the rest of kubelet is still running? 

Also this approach is consistent with the current volume manager. Please let me know what you think.

Reply:
> I think this is better than stopping the watcher if an error is encountered when traversing the directories because I don't see a way for the kubelet to restart the watcher again. Does it make sense for the watcher to stop if the rest of kubelet is still running?

That seems ok to me. Log an error and continue instead of completely giving up.

Looks like https://github.com/kubernetes/kubernetes/pull/75110 recently merged. Can you make sure this PR doesn't undo those changes. I don't know how relevant that bug is with this refactor, but lets minimize the unnecessary change unless we have a good reason for it.